### PR TITLE
Test train_dataset types at init for core trainers

### PR DESCRIPTION
**Pull request by albertvillanova**
This PR adds `test_init_with_train_dataset` to DPOTrainer, GRPOTrainer, KTOTrainer, RewardTrainer, RLOOTrainer and SFTTrainer, parametrized over `train_dataset` container types.

**Depends on**:
- [x] #6493

Until that PR merges, the `unsupported_dataset_dict` case added here has nothing to catch and CI is expected to fail on it. Merge order: #6493 first, then this one.

### Motivation

This is the next step (PR 2) in the small series aligning `train_dataset` handling and test coverage across core trainers with the `eval_dataset` work already done (#6492 and #6493 are the earlier steps). `eval_dataset` already has a dedicated, parametrized `test_init_with_eval_dataset` per trainer; `train_dataset` had no equivalent.

### Solution

Each new test is parametrized over `["dataset", "iterable_dataset", "none", "unsupported_dataset_dict"]`:

- `dataset` / `iterable_dataset`: construction succeeds; `trainer.train_dataset` is checked (tokenized column present for DPO/KTO/Reward/SFT, identity match for GRPO/RLOO, which pass `train_dataset` through untouched).
- `none`: raises the pre-existing `` ValueError("`train_dataset` is required") ``.
- `unsupported_dataset_dict`: wraps the dataset in a `DatasetDict` and expects the `TypeError` added in #6493.

The unsupported case is intentionally not exhaustive (no `dict_of_dataset`, no plain `list`, etc.): the underlying check is a single `isinstance` test, not per-type logic, so one representative unsupported container type is enough to confirm the guard fires.

### Changes

- `tests/test_dpo_trainer.py`: add `test_init_with_train_dataset`
- `tests/test_grpo_trainer.py`: add `test_init_with_train_dataset`
- `tests/test_kto_trainer.py`: add `test_init_with_train_dataset`
- `tests/test_reward_trainer.py`: add `test_init_with_train_dataset`
- `tests/test_rloo_trainer.py`: add `test_init_with_train_dataset`
- `tests/test_sft_trainer.py`: add `test_init_with_train_dataset`

---

> [!NOTE]
> **Low Risk**
> Test-only changes with no production code; CI for the unsupported-dataset case depends on merging #6493 first.
> 
> **Overview**
> Replaces the single **`test_train_dataset_required`** check on DPO, KTO, Reward, SFT, GRPO, and RLOO with **`test_init_with_train_dataset`**, mirroring the existing **`test_init_with_eval_dataset`** pattern.
> 
> Each new test is parametrized over **`dataset`**, **`iterable_dataset`** (with **`max_steps`** set), **`none`** (expects the existing **`ValueError`**), and **`unsupported_dataset_dict`** (expects **`TypeError`** from #6493). Success paths assert tokenized columns where trainers preprocess data, or identity on **`train_dataset`** for GRPO/RLOO.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit dbce85b4e53a5e8b59e3be089924a379932f462a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/test_rloo_trainer.py [hunk 1]
```diff
@@ -236,14 +236,49 @@ def test_train_with_iterable_dataset(self):
...
+                    train_dataset=train_dataset,
+                )
+        else:
+            trainer = RLOOTrainer(
                 model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
                 reward_funcs="trl-internal-testing/tiny-Qwen2ForSequenceClassification-2.5",
                 args=training_args,
+                train_dataset=train_dataset,
             )
+            assert trainer.train_dataset is train_dataset
```
Comment (qgallouedec) [inline tests/test_rloo_trainer.py:278]: nit, for grpo/rloo, this would probably be more consistent with the other tests:

```python
assert "prompt" in next(iter(trainer.train_dataset))
```

**Review by qgallouedec — APPROVED** [review]
lgtm!

**Comment by qgallouedec** [comment]
ah, CI is failing actually!

**Review by qgallouedec — CHANGES_REQUESTED** [review]
fix ci

**Review by qgallouedec — APPROVED** [review]
...because it needs https://github.com/huggingface/trl/pull/6493, now I remember.

File: tests/test_dpo_trainer.py [hunk 2]
```diff
@@ -1195,10 +1195,44 @@ def test_train_with_multiple_eval_dataset(self):
...
+            with pytest.raises(ValueError, match="`train_dataset` is required"):
+                DPOTrainer(
+                    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
+                    args=training_args,
+                    train_dataset=train_dataset,
+                )
+        elif train_dataset_type == "unsupported_dataset_dict":
+            with pytest.raises(
+                TypeError,
+                match="`train_dataset` doesn't support `DatasetDict`, `IterableDatasetDict`, or a `dict` of datasets",
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py]: depending on what you choose for https://github.com/huggingface/trl/pull/6493#issuecomment-5065860190, you could need to change these `match`

**Comment by albertvillanova** [comment]
The CI error is unrelated; see:
- #6532
